### ISSUE TITLE
Improve Adhoc filter

### ISSUE THREAD

**Issue by dsztykman** [grafana/clickhouse-datasource#120]
**Is your feature request related to a problem? Please describe.**
When you have a very large dataset spanning over multiple months/years, the adhoc query filter generates a single query:
`SELECT DISTINCT $column FROM $table LIMIT 1000`
This query can be very expensive as it can span over millions/billions of rows.
Also it'd be great to order the response by count or something else.

**Describe the solution you'd like**
We should add the timerange filter of the current dashboard into the adhoc filter.
So let's say your dashboard is doing now - 15 min the query should be something like:
`SELECT DISTINCT $column FROM $table WHERE $__timeFilter($timecolumn) LIMIT 1000`
Here the $__timeFilter would replace the query with the proper timerange.

Obviously you can't necessary know the name of the timefiltering column, so we should create a new constant variable like `clickhouse_adhoc_query_timecolumn` where user would specify the timecolumn to be used.

In general customising the query filter generated would be very interesting.
Another option would be to create a variable like called `$adhoc_filtered_column` which would contain the value of the column selected by the user.
And then let the user manage how the query filter is created with a constant variable `clickhouse_adhoc_query_filter` where user would use the `$adhoc_filtered_column` and craft their own filters like:
`SELECT DISTINCT $adhoc_filtered_column FROM table WHERE $__timeFilter(timecolumn) ORDER BY COUNT() LIMIT 10`

I prefer option 2 as the user can optimise and customise easily how they want to use adhocfilter

**Comment by gingerwizard** [comment]
Thanks @dsztykman 
We discussed this and it definitely seems like a good enhancement. I think we will expose the ability to control the specific query used to get the values i.e. set `SELECT DISTINCT $column FROM $table WHERE $__timeFilter($timecolumn) LIMIT 1000` 
`$column` will be available as a variable, as well as the normal global filters e.g. `$__timeFilter(columnName)`

The default will remain `SELECT DISTINCT $column FROM $table LIMIT 1000`

This will allow use of things like SAMPLE.

**Comment by bossinc** [comment]
@dsztykman I believe you are asking for a way to customize the adhoc filter. You can already customize the adhoc filter by creating a constant variable named `clickhouse_adhoc_query` with a value of your query. You can use the time range filter in your query.
https://grafana.com/grafana/plugins/grafana-clickhouse-datasource/#ad-hoc-filters

**Comment by dsztykman** [comment]
You can customise with a static value to specify which table you are working on.
But you can't customise how to retrieve the value out of a column that you dynamically select from the adhoc filter.
That's what I'm looking for here.

**Comment by gingerwizard** [comment]
We aren't clear what you mean @dsztykman 
You can now control the query used for ad-hoc filters - using both the $column and time variables. Are you looking for the query to be column dependent? i.e. different queries used for different columns?

**Comment by bossinc** [comment]
I implemented `column` to return all the columns that are being used for adhoc filters. 
In my screenshot, I have an adhoc filter for `machine_group` and `machine_name` and `$__adhocColumns` is returning these two columns.
![image](https://user-images.githubusercontent.com/5140848/175994238-d9ad0279-4dfe-470e-af53-0fd3adc981c4.png)
@dsztykman I am pretty sure this isn't what you are imagining this variable to do. What value is expected to be returned by column?

**Comment by dsztykman** [comment]
No it's definitely not what I had in mind.
Basically I want to be able to specify a query filter in the adhoc filters.
Let's say you have a table called `sample.cts` and you want to enable adhoc filtering on this table in particular but in that table you want to limit adhoc query to the current timerange and order by count desc.

```SELECT $column FROM sample.cts WHERE $__timeFilter(timestamp) ORDER BY COUNT() DESC GROUP BY $column LIMIT 100```

This will populate the different value in the adhoc filter, based on the column selected by the users.
The `clickhouse_adhoc_query` allow you to control the column selection whereas I want to filter the value after selecting the column.
(Does that make sense?)

**Comment by gingerwizard** [comment]
Maybe we should change the name `clickhouse_adhoc_query` to `clickhouse_adhoc_column_query` and introduce `clickhouse_adhoc_value_query`..would this fulfill the requirement @dsztykman ?

**Comment by dsztykman** [comment]
Yeah that would be good, but I would also need a variable for the selected column, so I can I create a constant variable called `clickhouse_adhoc_value_query` which would contain:
```SELECT DISTINCT $column_adhoc FROM table ORDER BY COUNT() LIMIT 100```
For example

**Comment by gingerwizard** [comment]
Sorry im confused `clickhouse_adhoc_query` currently allows you to control both the column and values. Are you looking to be able to specify a different query for populating the values PER column?

**Comment by dsztykman** [comment]
I think I got confused :)
Technically yes I think `clickhouse_adhoc_column_query` and `clickhouse_adhoc_value_query` would fulfill the requirement.
Let's say I have 30 different tables, with the `clickhouse_adhoc_column_query` I could specify the table I want to use to get the column list.
Then with `clickhouse_adhoc_value_query` I could specify how to retrieve the list of values.
Does that make sense?

**Comment by gingerwizard** [comment]
I discussed this with @dsztykman, and i think there are two requests here:

1. Currently `clickhouse_adhoc_query` returns the column/s and values to render for selection. You can return multiple columns and values. This is limited in that one query will control the values for all columns, e.g. you can't have a different limit or ordering per column - per SQL limitations. Columns from multiple tables are possible but you'd need to use a JOIN/UNION and this is restrictive and likely slow. 
2. (1) means that the same date field is available for all columns. This is currently likely as all queries will be from the same table. However, for multiple table scenarios, this might not be the case.
3. One option might be a different ad-hoc filter table. Unfortunately, this isn't possible. The `clickhouse_adhoc_query` applies to all ad-hoc filters. Currently, if you have multiple ad-hoc filters on a dashboard for different tables, the same `clickhouse_adhoc_query` value will be used.

Allowing multiple ad-hoc filters, each with their own `clickhouse_adhoc_query`, would address alot of these issues.

1. For multiple tables, users could simply create separate ad-hoc filters associating each with their own filter.
2. If users need separate queries for each column, they could in theory group these into different ad-hoc queries.

This is imperfect since (2) might not always be viable but having a query for each column isn't better IMO - it would still require creation of lots of constant variables per column.

**Comment by gingerwizard** [comment]
Discussed with @bossinc 

The way we think about this is template variables give you full context of the column - allowing you to specify a query per column. With hundreds of columns, however, this results in hundreds of drop-downs. This isn't realistic.

The other extreme (which is the current implementation of ad-hoc filters) is no context of the column. This means a single query for all columns - column values cant have a different ORDER and we can't filter them by a different date field.

The compromise, especially for column-orientated dbs with many columns, is probably an ad-hoc filter which is table aware. Discussing with @bossinc this feels like a core Grafana request. Most columns in the same table would be ordered and filtered by the same time column for example.

@dsztykman

**Comment by bossinc** [comment]
Closing. Please reopen if there is more to discuss.

**Comment by dsztykman** [comment]
@bossinc @gingerwizard sorry to come back to that subject again after awhile.
But unfortunately the current implementation doesn't really workout for my problem.

A quick example, I have data for more than 9 months which are using lots of TB.
Currently the adhoc filter for this example:
![Screenshot 2023-03-06 at 14 26 46](https://user-images.githubusercontent.com/33227713/223123136-f19e6f9c-3b96-4732-8147-bc4f01a48935.png)

Is generating the following query:

```
select distinct UA from akamai.logs limit 1000
```
Which is a very very large dataset for me, to improve the performance I would like to limit this query to run only on the current dashboard timeframe, so being able to add a WHERE statement including:
```
WHERE $__timeFilter(reqTimeSec)
```
(reqTimeSec is my primary timestamp I'm using in Grafana to fiter the data).

Does that make more sense?
Basically having the ability to customize the adhoc filter value query to add a WHERE condition to limit the timerange to get the value.

**Comment by fredmnl** [comment]
Hi everyone!

I still don't understand how `clickhouse_adhoc_query` can be used to create anything more complex than operating on a single column. I'm looking to enable adhoc filtering on only a few columns of my table (which has too many columns to be a practical dropdown menu), say approximately 5 out of 50.

I have tried `clickhouse_adhoc_query` : `SELECT DISTINCT col1 AS readable_name, col2, col3 FROM db.table` but this results in duplicated values for all columns as they will appear multiple times in the resulting query (as they are indeed part of multiple DISTINCT sets). It does however provide me with my desired subset of columns and a nice way to rename them. 

I have tried template variables as you suggested @gingerwizard. They allow very nice flexibility in the way we query the columns but I don't see how I could be using the result from such query in an adhoc filter. Am I missing something there? I don't want to modify my SQL queries to handle the template variables (and especially the cases where the variable is not selected) and that's kind of the promise of the adhoc filters. 

I hope reopening this ticket is the right move but this feels right along the discussion above. 

Thanks!
